feat(Input): make baseinput field controlled - PR2

kamleshchandnani opened the pull request:
This PR adds the following capabilities to the input field
- [x] capability to make the field controlled
- [x] capability to make the field isDisabled
- [x] label for input field
- [x] handle validations on input field
- [x] capability for isRequired

[API Decisions Doc](https://github.com/razorpay/blade/blob/master/packages/blade/src/components/Input/_decisions/decisions.md)

Previous PR link for some context
PR1 - https://github.com/razorpay/blade/pull/602

Remaining capabilities
PR3 - #632

anuraghazra commented on packages/blade/src/components/Box/getBoxStyles.ts:
```diff
@@ -4,14 +4,17 @@ import type { BoxProps } from './Box';
 
 // allows 'auto' as spacing value
-const getAutoSpacingValue = <T extends string | undefined>({
+const getAutoSpacingValue = <T extends string | number | undefined>({
```
Let's rename this function to "getSpacingValue" then.

anuraghazra commented on packages/blade/src/components/Checkbox/useCheckbox.ts:
```diff
@@ -5,7 +5,7 @@ import type { GestureResponderEvent } from 'react-native';
 import { getPlatformType, makeAccessible } from '../../utils';
 import type { CheckboxProps } from './Checkbox';
-import { useFormId } from '~src/hooks/useFormId';
+import { useFormId } from '~components/Form/useFormId';
```
Why put hooks in Form folder?

anuraghazra replied:
Oh I see. the useFormId hook. I guess makes sense then.

anuraghazra commented on packages/blade/src/components/Input/BaseInput/BaseInput.tsx:
```diff
@@ -1,73 +1,166 @@
-  name?: string;
-  value?: string;
-}) => void;
+export type OnChange = ({ name, value }: { name?: string; value?: string }) => void;
```
event?

kamleshchandnani replied:
don't need event in case of input field. just sending out name and value to the consumer. also event is diff on react native and essentially there's no use as of now

anuraghazra commented on packages/blade/src/components/Input/BaseInput/BaseInput.tsx:
```diff
@@ -1,73 +1,166 @@
+  }
+
+  const handleOnChange: HandleOnChange = React.useCallback(
+    ({ inputName, inputValue }) => {
```
Why did we called it `inputName` `inputValue` instead of just `name` `value`

kamleshchandnani replied:
scope clash 😢 
trying to fix it now since a lot of changes have gone through let me check

anuraghazra commented on packages/blade/src/components/Input/BaseInput/BaseInput.tsx:
```diff
@@ -1,73 +1,166 @@
+    ({ inputName, inputValue }) => {
+      let _inputValue = '';
+
+      if (getPlatformType() === 'react-native' && typeof inputValue === 'string') {
+        _inputValue = inputValue;
+      } else if (typeof inputValue !== 'string') {
+        // it's weird but TS forced me to write this much code where I could have just done "getPlatformType() === 'react-native' ? inputValue : inputValue?.target.value" :(
+        _inputValue = inputValue?.target.value ?? '';
+      }
```
Does this look a bit more cleaner? 

> also uplift the `const isReactNative = getPlatformType() === 'react-native';` check to component body, no need to check on each and every keystroke. 

```ts
      const isValueString = typeof inputValue === 'string';
      if (isReactNative && isValueString) {
        _inputValue = inputValue;
      }
      if (!isValueString) {
        // it's weird but TS forced me to write this much code where I could have just done "getPlatformType() === 'react-native' ? inputValue : inputValue?.target.value" :(
        _inputValue = inputValue?.target.value ?? '';
      }
```

kamleshchandnani replied:
don't mind doing that but honestly i don't see any value of doing that. it's just moving things around 1 level up where we don't even get any other advantage. if `getPlatformType` and `typeof` operation would have been very heavy i would have done that but they are very very very light weight operations so wouldn't make any diff by moving up at the cost of readability.

anuraghazra commented on packages/blade/src/components/Input/BaseInput/BaseInput.tsx:
```diff
@@ -1,73 +1,166 @@
+  _validationState,
+  hasHelpText,
+}: {
+  _validationState: BaseInputProps['validationState'];
```
No need of _ prefix

kamleshchandnani replied:
yeah some spaghetti code, fixed

anuraghazra commented on packages/blade/src/components/Input/BaseInput/StyledBaseInput.native.tsx:
```diff
@@ -1,49 +1,39 @@
+  handleOnChange,
+  ...props
+}: StyledBaseInputProps): ReactElement => {
+  const [isFocused, setisFocused] = useState(false);
```
We can utilize our useInteraction hook. 

```ts
  const { currentInteraction, onBlur, onFocus } = useInteraction();
  const isFocused = currentInteraction === 'focus';
```

kamleshchandnani replied:
let me check that

kamleshchandnani replied:
what is the benefit of this useInteraction?

anuraghazra commented on packages/blade/src/components/Input/BaseInput/getBaseInputStyles.ts:
```diff
@@ -1,26 +1,74 @@
+  isFocused,
+  isDisabled,
+  validationState,
+}: Pick<BaseInputProps, 'isDisabled' | 'validationState'> & {
+  isFocused?: boolean;
+  theme: Theme;
+}): CSSObject => {
```
Let's not inline complex types on function body, looks very cluttered. 

Also `getInputBackgroundAndBorderStyles` & `getBaseInputStyles` has the exact same type defs, check if we can just reuse it.

```ts
type GetBaseInputStylesProps = Pick<BaseInputProps, 'isDisabled' | 'validationState'> & {
  isFocused?: boolean;
  theme: Theme;
};
```

anuraghazra commented on packages/blade/src/components/Input/BaseInput/BaseInput.tsx:
```diff
@@ -77,26 +170,60 @@ export const BaseInput = ({
+        <FormLabel
+          as="label"
+          neccessityIndicator={neccessityIndicator}
+          id={labelId}
```
Oh actually ID isn't needed when using it `as=label` can you remove it and make the ID prop on FormLabel optional.

kamleshchandnani replied:
it'll be needed right? because we are not wrapping the text input inside label so we need to associate the label with the input id

anuraghazra replied:
label doesn't need ID. it only needs htmlFor. 

input needs ID.

kamleshchandnani replied:
okay i confused these 2 props.

anuraghazra commented on packages/blade/src/components/Input/BaseInput/BaseInput.tsx:
```diff
@@ -77,26 +170,60 @@ export const BaseInput = ({
+      <Box marginLeft={labelPosition === 'left' ? 120 : 'auto'}>
+        <FormHint
+          type={getHintType({ _validationState: validationState, hasHelpText: Boolean(helpText) })}
+          helpText={helpText}
+          errorText={errorText}
+          successText={successText}
+          helpTextId={helpTextId}
+          errorTextId={errorTextId}
+          successTextId={successTextId}
```
> TODO: write up a11y ids to inputs

kamleshchandnani replied:
A11y is next milestone - PR4 maybe

anuraghazra commented on packages/blade/src/components/Input/BaseInput/BaseInput.tsx:
```diff
@@ -1,73 +1,166 @@
+   * If `error`, the input is marked as invalid,
+   * and `invalid` attribute will be added
+   *
+   * If `success`, the input is marked as invalid,
```
If success marked as invalid?

kamleshchandnani replied:
typo

anuraghazra commented on packages/blade/src/components/Form/FormLabel.tsx:
```diff
@@ -112,7 +118,7 @@ const FormLabel = ({
   return (
-    <Component htmlFor={htmlFor} style={{ width }} id={id}>
+    <Component htmlFor={htmlFor} style={{ width, flexShrink: 0 }} id={id}>
       <Box marginBottom="spacing.1">{textNode}</Box>
```
Oops actually I forgot to do this: 

The margin bottom should only apply when position is top: 

```ts
position === 'left' ? 'spacing.0' : 'spacing.1'
```

kamleshchandnani replied:
ideally shouldn't matter but fine i'll fix it

kamleshchandnani commented:
@anuraghazra addressed all your comments, can you re-review and then change the label to L2 once you're done? @divyanshu013 tagged you for L2 review which you can start in parallel

divyanshu013 commented on packages/blade/src/components/Box/getBoxStyles.ts:
```diff
@@ -4,14 +4,17 @@ import type { BoxProps } from './Box';
 
 // allows 'auto' as spacing value
-const getAutoSpacingValue = <T extends string | undefined>({
+const getSpacingValue = <T extends string | number | undefined>({
```
Why are we allowing numbers? Won't that break the purpose of tokens 🤔 
If there's a legit usecase we should add a comment to explain the intent

chaitanyadeorukhkar replied:
+1. Now that we don't need all the "absolute" hacks. Can we get avoid allowing free-form margins?

kamleshchandnani replied:
will fix this in Pr-3 since the everything resides there for this so will check

divyanshu013 commented on packages/blade/src/components/Form/FormLabel.tsx:
```diff
@@ -108,12 +114,13 @@ const FormLabel = ({
+  const isLabelLeftPositioned = position === 'left';
   // only set 120px label when device is desktop
-  const width = position === 'left' && isDesktop ? '120px' : 'auto';
+  const width = isLabelLeftPositioned && isDesktop ? '120px' : 'auto';
```
Optional: I think we should move this width to somewhere on the top as a `SOME_CONSTANT` for ease of readability.

chaitanyadeorukhkar replied:
We should have a `formTokens.js` where we can store this. Doing the same for other components where with fixed dimensions

kamleshchandnani replied:
it's just one tiny thing don't think that creating a whole file for just this will add any value.

divyanshu013 commented on packages/blade/src/components/Input/BaseInput/BaseInput.tsx:
```diff
@@ -1,73 +1,167 @@
+    ({ name, value }) => {
+      let _value = '';
+
+      if (getPlatformType() === 'react-native' && typeof value === 'string') {
```
Can we move platform specific logic to .web / .native files?

kamleshchandnani replied:
so at every step you kind of ask this question that shall we move this to a new file or same file with a basic check. In this case it's just 2-3 things which are diff rest the entire thing is same so putting a check is less noisy and less duplication

divyanshu013 reviewed (commented):
Looks good overall, minor comments.

chaitanyadeorukhkar commented:
Cant render this PR on react native @kamleshchandnani 
<img width="200" alt="Screenshot 2022-08-23 at 4 53 17 PM" src="https://user-images.githubusercontent.com/24487274/186145907-dadc8daa-e0b4-4dad-9c66-779627bc5ff1.png">

divyanshu013 reviewed (approved):
Looks good to me.

kamleshchandnani commented:
han could be merge conflicts after icon thing 😢 

fixing it now

> Cant render this PR on react native @kamleshchandnani <img alt="Screenshot 2022-08-23 at 4 53 17 PM" width="200" src="https://user-images.githubusercontent.com/24487274/186145907-dadc8daa-e0b4-4dad-9c66-779627bc5ff1.png">
